File: packages/workflow/system-nexus/typescript/temporal_model_converters.ts
import * as common from "@temporalio/common";
import type { google, temporal } from "@temporalio/proto";
import * as workflow from "@temporalio/workflow";
import type Long from "long";

function int64ToNumber(
  value: Long | number | string | object | null | undefined,
): number {
  if (value == null) {
    return 0;
  }
  if (typeof value === "number") {
    return value;
  }
  if (typeof value === "string") {
    return Number(value);
  }
  if ("toNumber" in value && typeof value.toNumber === "function") {
    return value.toNumber();
  }
  if ("low" in value && "high" in value) {
    const longValue = value as {
      low: number;
      high: number;
      unsigned?: boolean;
    };
    const low = longValue.low >>> 0;
    return longValue.high * 4_294_967_296 + low;
  }
  throw new TypeError("unsupported int64 value");
}

function durationToMillis(
  proto: google.protobuf.IDuration | null | undefined,
): number | undefined {
  if (proto == null) {
    return undefined;
  }
  return (
    int64ToNumber(proto.seconds) * 1000 +
    Math.floor((proto.nanos ?? 0) / 1_000_000)
  );
}

export function retryPolicyFromProto(
  proto: temporal.api.common.v1.IRetryPolicy,
): common.RetryPolicy {
  return {
    backoffCoefficient: proto.backoffCoefficient ?? undefined,
    maximumAttempts: proto.maximumAttempts ?? undefined,
    maximumInterval: durationToMillis(proto.maximumInterval),
    initialInterval: durationToMillis(proto.initialInterval),
    nonRetryableErrorTypes: proto.nonRetryableErrorTypes ?? undefined,
  };
}

export function retryPolicyToProto(
  retryPolicy: common.RetryPolicy,
): temporal.api.common.v1.IRetryPolicy {
  return common.compileRetryPolicy(retryPolicy);
}

export function workflowTypeFromProto(
  proto: temporal.api.common.v1.IWorkflowType,
): string | common.Workflow {
  return proto.name ?? "";
}

export function workflowTypeToProto(
  workflowType: string | common.Workflow,
): temporal.api.common.v1.IWorkflowType {
  return { name: workflowFunctionName(workflowType) };
}

export function workflowFunctionName(value: string | common.Workflow): string {
  return typeof value === "string" ? value : common.extractWorkflowType(value);
}

export function signalFunctionToProto(
  value: string | workflow.SignalDefinition<any[]>,
): string {
  return typeof value === "string" ? value : value.name;
}

export function taskQueueFromProto(
  proto: temporal.api.taskqueue.v1.ITaskQueue,
): string {
  return proto.name ?? "";
}

export function taskQueueToProto(
  taskQueue: string,
): temporal.api.taskqueue.v1.ITaskQueue {
  return { name: taskQueue };
}

export function workflowNamespace(): string {
  return workflow.workflowInfo().namespace;
}

export function payloadFromProto(
  payload: temporal.api.common.v1.IPayload,
): common.Payload {
  return payload;
}

export function payloadToProto(
  payload: common.Payload,
): temporal.api.common.v1.IPayload {
  return payload;
}

function configuredPayloadConverter(): common.PayloadConverter {
  const activator = (
    globalThis as typeof globalThis & {
      __TEMPORAL_ACTIVATOR__?: {
        payloadConverter?: common.PayloadConverter;
      };
    }
  ).__TEMPORAL_ACTIVATOR__;
  if (activator?.payloadConverter == null) {
    throw new Error(
      "payload converter is unavailable outside workflow context",
    );
  }
  return activator.payloadConverter;
}

export function memoFromProto(
  proto: temporal.api.common.v1.IMemo,
): Record<string, unknown> {
  return (
    common.mapFromPayloads(
      configuredPayloadConverter(),
      proto.fields ?? undefined,
    ) ?? {}
  );
}

export function memoToProto(
  memo: Record<string, unknown>,
): temporal.api.common.v1.IMemo {
  return {
    fields: common.mapToPayloads(configuredPayloadConverter(), memo),
  };
}

export function durationFromProto(
  proto: google.protobuf.IDuration,
): common.Duration {
  return durationToMillis(proto)!;
}

export function durationToProto(
  duration: common.Duration,
): google.protobuf.IDuration {
  return common.msToTs(duration);
}

function typedSearchAttributePayload(
  value: unknown,
  type: common.SearchAttributeType,
): common.Payload {
  const payload = configuredPayloadConverter().toPayload(value);
  payload.metadata ??= {};
  payload.metadata.type = common.u8(
    common.TypedSearchAttributes.toMetadataType(type),
  );
  return payload;
}

function isValidSearchAttributeValue(
  type: common.SearchAttributeType,
  value: unknown,
): boolean {
  switch (type) {
    case common.SearchAttributeType.TEXT:
    case common.SearchAttributeType.KEYWORD:
      return typeof value === "string";
    case common.SearchAttributeType.INT:
      return Number.isInteger(value);
    case common.SearchAttributeType.DOUBLE:
      return typeof value === "number";
    case common.SearchAttributeType.BOOL:
      return typeof value === "boolean";
    case common.SearchAttributeType.DATETIME:
      return value instanceof Date;
    case common.SearchAttributeType.KEYWORD_LIST:
      return (
        Array.isArray(value) && value.every((item) => typeof item === "string")
      );
    default:
      return false;
  }
}

function typedSearchAttributePairFromPayload(
  name: string,
  payload: common.Payload,
): common.SearchAttributePair | undefined {
  const metadataType = payload.metadata?.type;
  if (metadataType == null) {
    return undefined;
  }
  const type = common.TypedSearchAttributes.toSearchAttributeType(
    common.str(metadataType),
  );
  if (type == null) {
    return undefined;
  }
  let value: unknown = configuredPayloadConverter().fromPayload(payload);
  if (
    type !== common.SearchAttributeType.KEYWORD_LIST &&
    Array.isArray(value)
  ) {
    if (value.length !== 1) {
      return undefined;
    }
    value = value[0];
  }
  if (type === common.SearchAttributeType.DATETIME && value != null) {
    value = new Date(value as string);
  }
  if (!isValidSearchAttributeValue(type, value)) {
    return undefined;
  }
  return {
    key: { name, type },
    value,
  } as common.SearchAttributePair;
}

export function searchAttributesFromProto(
  proto: temporal.api.common.v1.ISearchAttributes,
): common.TypedSearchAttributes {
  const indexedFields = proto.indexedFields ?? {};
  const typedPairs: common.SearchAttributePair[] = [];
  for (const [name, payload] of Object.entries(indexedFields)) {
    const pair = typedSearchAttributePairFromPayload(name, payload);
    if (pair == null) {
      throw new TypeError(
        `search attribute ${name} cannot be decoded as a typed search attribute`,
      );
    }
    typedPairs.push(pair);
  }
  return new common.TypedSearchAttributes(typedPairs);
}

export function searchAttributesToProto(
  searchAttributes: common.TypedSearchAttributes,
): temporal.api.common.v1.ISearchAttributes {
  return {
    indexedFields: Object.fromEntries(
      searchAttributes
        .getAll()
        .map((pair): [string, common.Payload] => [
          pair.key.name,
          typedSearchAttributePayload(pair.value, pair.key.type),
        ]),
    ),
  };
}

export function priorityFromProto(
  proto: temporal.api.common.v1.IPriority,
): common.Priority {
  return common.decodePriority(proto);
}

export function priorityToProto(
  priority: common.Priority,
): temporal.api.common.v1.IPriority {
  return common.compilePriority(priority);
}

const VERSIONING_BEHAVIOR_PINNED = 1;
const VERSIONING_BEHAVIOR_AUTO_UPGRADE = 2;
const PINNED_OVERRIDE_BEHAVIOR_PINNED = 1;

export function versioningOverrideFromProto(
  proto: temporal.api.workflow.v1.IVersioningOverride,
): common.VersioningOverride | undefined {
  if (
    proto.autoUpgrade ||
    proto.behavior === VERSIONING_BEHAVIOR_AUTO_UPGRADE
  ) {
    return "AUTO_UPGRADE";
  }
  const pinnedVersion = proto.pinned?.version;
  if (pinnedVersion?.deploymentName != null && pinnedVersion.buildId != null) {
    return {
      pinnedTo: {
        deploymentName: pinnedVersion.deploymentName,
        buildId: pinnedVersion.buildId,
      },
    };
  }
  if (
    proto.deployment?.seriesName != null &&
    proto.deployment.buildId != null
  ) {
    return {
      pinnedTo: {
        deploymentName: proto.deployment.seriesName,
        buildId: proto.deployment.buildId,
      },
    };
  }
  return undefined;
}

export function versioningOverrideToProto(
  versioningOverride: common.VersioningOverride,
): temporal.api.workflow.v1.IVersioningOverride {
  if (versioningOverride === "AUTO_UPGRADE") {
    return {
      behavior: VERSIONING_BEHAVIOR_AUTO_UPGRADE,
      autoUpgrade: true,
    };
  }
  return {
    behavior: VERSIONING_BEHAVIOR_PINNED,
    pinnedVersion: common.toCanonicalString(versioningOverride.pinnedTo),
    pinned: {
      behavior: PINNED_OVERRIDE_BEHAVIOR_PINNED,
      version: {
        deploymentName: versioningOverride.pinnedTo.deploymentName,
        buildId: versioningOverride.pinnedTo.buildId,
      },
    },
  };
}

export function workflowIdReusePolicyFromProto(
  policy: temporal.api.enums.v1.WorkflowIdReusePolicy,
): common.WorkflowIdReusePolicy | undefined {
  return common.decodeWorkflowIdReusePolicy(policy);
}

export function workflowIdReusePolicyToProto(
  policy: common.WorkflowIdReusePolicy,
): temporal.api.enums.v1.WorkflowIdReusePolicy | undefined {
  return common.encodeWorkflowIdReusePolicy(policy);
}

export function workflowIdConflictPolicyFromProto(
  policy: temporal.api.enums.v1.WorkflowIdConflictPolicy,
): common.WorkflowIdConflictPolicy | undefined {
  return common.decodeWorkflowIdConflictPolicy(policy);
}

export function workflowIdConflictPolicyToProto(
  policy: common.WorkflowIdConflictPolicy,
): temporal.api.enums.v1.WorkflowIdConflictPolicy | undefined {
  return common.encodeWorkflowIdConflictPolicy(policy);
}

File: packages/workflow/system-nexus-wit/deps/nexus-temporal-types/python/temporal_model_converters.py
import collections.abc
from datetime import timedelta
import typing

import google.protobuf.duration_pb2
import temporalio.api.common.v1.message_pb2 as common_pb2
import temporalio.api.enums.v1.workflow_pb2 as workflow_enums_pb2
import temporalio.api.failure.v1.message_pb2 as failure_pb2
import temporalio.api.taskqueue.v1.message_pb2 as taskqueue_pb2
import temporalio.api.workflow.v1 as workflow_pb2
import temporalio.converter as temporalio_converter
import temporalio.common as temporalio_common
import temporalio.nexus.system
import temporalio.exceptions as temporalio_exceptions


class SignalWithStartWorkflowModelRequest(typing.Protocol):
    namespace: str
    id: str


def retry_policy_from_proto(
    proto: common_pb2.RetryPolicy,
) -> temporalio_common.RetryPolicy:
    return temporalio_common.RetryPolicy.from_proto(proto)


def retry_policy_to_proto(
    retry_policy: temporalio_common.RetryPolicy,
) -> common_pb2.RetryPolicy:
    proto = common_pb2.RetryPolicy()
    retry_policy.apply_to_proto(proto)
    return proto


def workflow_function_name(
    value: str | collections.abc.Callable[..., collections.abc.Awaitable[object]],
) -> str:
    from temporalio.workflow import _Definition

    name, _result_type = _Definition.get_name_and_result_type(value)
    return name


def signal_function_to_proto(
    value: str | collections.abc.Callable[..., typing.Any],
) -> str:
    from temporalio.workflow import _SignalDefinition

    return _SignalDefinition.must_name_from_fn_or_str(value)


def workflow_type_to_proto(
    workflow_type: str
    | collections.abc.Callable[..., collections.abc.Awaitable[object]],
) -> common_pb2.WorkflowType:
    return common_pb2.WorkflowType(name=workflow_function_name(workflow_type))


def workflow_type_from_proto(
    proto: common_pb2.WorkflowType,
) -> str:
    return proto.name


def task_queue_from_proto(
    proto: taskqueue_pb2.TaskQueue,
) -> str:
    return proto.name


def task_queue_to_proto(
    task_queue: str,
) -> taskqueue_pb2.TaskQueue:
    return taskqueue_pb2.TaskQueue(name=task_queue)


def workflow_namespace() -> str:
    from temporalio.workflow import info

    return info().namespace


def signal_with_start_workflow_serialization_context(
    request: SignalWithStartWorkflowModelRequest,
) -> temporalio_converter.WorkflowSerializationContext:
    return temporalio_converter.WorkflowSerializationContext(
        namespace=request.namespace,
        workflow_id=request.id,
    )


def payloads_to_proto(
    values: collections.abc.Sequence[typing.Any],
) -> common_pb2.Payloads:
    return temporalio.nexus.system._current_user_payload_converter().to_payloads_wrapper(values)


def payloads_from_proto(
    proto: common_pb2.Payloads,
    type_hints: list[typing.Any] | None = None,
) -> list[typing.Any]:
    if not proto.payloads:
        return []
    return temporalio.nexus.system._current_user_payload_converter().from_payloads(
        proto.payloads,
        type_hints,
    )


def _clone_payload(payload: common_pb2.Payload) -> common_pb2.Payload:
    clone = common_pb2.Payload()
    clone.CopyFrom(payload)
    return clone


def _value_to_payload(
    value: object | common_pb2.Payload,
) -> common_pb2.Payload:
    if isinstance(value, common_pb2.Payload):
        return _clone_payload(value)

    payloads = temporalio.nexus.system._current_user_payload_converter().to_payloads_wrapper([value])
    return _clone_payload(payloads.payloads[0])


def _payload_to_value(
    payload: common_pb2.Payload,
) -> object:
    wrapper = common_pb2.Payloads()
    wrapper.payloads.add().CopyFrom(payload)

    return typing.cast(
        object,
        temporalio.nexus.system._current_user_payload_converter().from_payloads_wrapper(wrapper)[0],
    )


_PayloadT = typing.TypeVar("_PayloadT")


@typing.overload
def payload_from_proto(
    proto: common_pb2.Payload,
    type_hint: type[_PayloadT],
) -> _PayloadT: ...


@typing.overload
def payload_from_proto(
    proto: common_pb2.Payload,
    type_hint: None = None,
) -> typing.Any: ...


def payload_from_proto(
    proto: common_pb2.Payload,
    type_hint: type[typing.Any] | None = None,
) -> typing.Any:
    converter = temporalio.nexus.system._current_user_payload_converter()
    if type_hint is None:
        return converter.from_payload(_clone_payload(proto))
    return converter.from_payload(_clone_payload(proto), type_hint)


def payload_to_proto(
    payload: object,
) -> common_pb2.Payload:
    return _value_to_payload(payload)


def failure_from_proto(
    proto: failure_pb2.Failure,
) -> BaseException:
    return temporalio_converter.FailureConverter.default.from_failure(
        proto,
        _failure_payload_converter(),
    )


def failure_to_proto(
    failure: BaseException,
) -> failure_pb2.Failure:
    proto = failure_pb2.Failure()
    temporalio_converter.FailureConverter.default.to_failure(
        failure,
        _failure_payload_converter(),
        proto,
    )
    return proto


def _failure_payload_converter() -> temporalio_converter.PayloadConverter:
    try:
        return temporalio.nexus.system._current_user_payload_converter()
    except RuntimeError:
        try:
            from temporalio.workflow import payload_converter

            return payload_converter()
        except temporalio_exceptions.TemporalError:
            return temporalio_converter.PayloadConverter.default


def memo_from_proto(
    proto: common_pb2.Memo,
) -> collections.abc.Mapping[str, object]:
    return {key: _payload_to_value(value) for key, value in proto.fields.items()}


def memo_to_proto(
    memo: collections.abc.Mapping[str, object],
) -> common_pb2.Memo:
    message = common_pb2.Memo()
    for key, value in memo.items():
        message.fields[key].CopyFrom(_value_to_payload(value))
    return message


def header_from_proto(
    proto: common_pb2.Header,
) -> collections.abc.Mapping[str, object]:
    return {key: _payload_to_value(value) for key, value in proto.fields.items()}


def header_to_proto(
    header: collections.abc.Mapping[str, object],
) -> common_pb2.Header:
    message = common_pb2.Header()
    for key, value in header.items():
        message.fields[key].CopyFrom(_value_to_payload(value))
    return message


def duration_from_proto(
    proto: google.protobuf.duration_pb2.Duration,
) -> timedelta:
    return proto.ToTimedelta()


def duration_to_proto(
    duration: timedelta,
) -> google.protobuf.duration_pb2.Duration:
    proto = google.protobuf.duration_pb2.Duration()
    proto.FromTimedelta(duration)
    return proto


def workflow_id_reuse_policy_from_proto(
    policy: workflow_enums_pb2.WorkflowIdReusePolicy.ValueType,
) -> temporalio_common.WorkflowIDReusePolicy:
    return temporalio_common.WorkflowIDReusePolicy(int(policy))


def workflow_id_reuse_policy_to_proto(
    policy: temporalio_common.WorkflowIDReusePolicy,
) -> workflow_enums_pb2.WorkflowIdReusePolicy.ValueType:
    return typing.cast(workflow_enums_pb2.WorkflowIdReusePolicy.ValueType, int(policy))


def workflow_id_conflict_policy_from_proto(
    policy: workflow_enums_pb2.WorkflowIdConflictPolicy.ValueType,
) -> temporalio_common.WorkflowIDConflictPolicy:
    return temporalio_common.WorkflowIDConflictPolicy(int(policy))


def workflow_id_conflict_policy_to_proto(
    policy: temporalio_common.WorkflowIDConflictPolicy,
) -> workflow_enums_pb2.WorkflowIdConflictPolicy.ValueType:
    return typing.cast(
        workflow_enums_pb2.WorkflowIdConflictPolicy.ValueType, int(policy)
    )


def search_attributes_to_proto(
    search_attributes: temporalio_common.TypedSearchAttributes,
) -> common_pb2.SearchAttributes:
    proto = common_pb2.SearchAttributes()
    temporalio_converter.encode_search_attributes(search_attributes, proto)
    return proto


def search_attributes_from_proto(
    proto: common_pb2.SearchAttributes,
) -> temporalio_common.TypedSearchAttributes:
    return temporalio_converter.decode_typed_search_attributes(proto)


def priority_from_proto(
    proto: common_pb2.Priority,
) -> temporalio_common.Priority:
    return temporalio_common.Priority._from_proto(proto)


def priority_to_proto(
    priority: temporalio_common.Priority,
) -> common_pb2.Priority:
    return priority._to_proto()


def versioning_override_to_proto(
    versioning_override: temporalio_common.VersioningOverride,
) -> workflow_pb2.VersioningOverride:
    return versioning_override._to_proto()


def versioning_override_from_proto(
    proto: workflow_pb2.VersioningOverride,
) -> temporalio_common.VersioningOverride:
    if proto.HasField("pinned") and proto.pinned.HasField("version"):
        version = proto.pinned.version
        return temporalio_common.PinnedVersioningOverride(
            temporalio_common.WorkerDeploymentVersion(
                deployment_name=version.deployment_name,
                build_id=version.build_id,
            )
        )
    if proto.pinned_version:
        return temporalio_common.PinnedVersioningOverride(
            temporalio_common.WorkerDeploymentVersion.from_canonical_string(
                proto.pinned_version
            )
        )
    if proto.auto_upgrade:
        return temporalio_common.AutoUpgradeVersioningOverride()
    raise ValueError("unknown versioning override proto shape")
